### PR TITLE
feat: unitinput

### PR DESCRIPTION
**Pull request by lisalupi**
## Type

- Feature

#### What is expected?

New component `<UnitInput />`!
Also, "disabled" is now optional for the options in <SelectInputV2 />
#### The following changes where made:

1. New component with tests + stories

2. Optional prop `disabled` for the options in selectinputV2

### CODE REVIEW

File: packages/ui/src/components/UnitInput/index.tsx [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+type UnitInputValue = { inputValue: number; unit: string }
+
+type UnitInputProps = {
+  id?: string
+  className?: string
+  name: string
+  disabled?: boolean
+  maxValue?: number
+  minValue?: number
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:141]: You can name it `max` and `min` in order to follow same naming as `NumberInputV2`

File: packages/ui/src/components/UnitInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+  width?: ComponentProps<typeof Stack>['width']
+  placeholderUnit?: string
+  autoFocus?: boolean
+  error?: boolean
+  success?: boolean
+  label?: string
+  labelInformation?: ReactNode
+  step?: number | string
+  readOnly?: boolean
+}
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:164]: For the following props: name, id, disabled, placeholder, autoFocus, required, and if you find more. You can do:

```ts
} & InputHTMLAttributes<HTMLInputElement>,
  | 'name'
  | 'id'
  | 'placeholder'
  | ...
```

File: packages/ui/src/components/UnitInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+  value?: UnitInputValue['inputValue']
+  unitValue?: UnitInputValue['unit']
+  onChange: (value: UnitInputValue['inputValue']) => void
+  onChangeUnitValue: (values: string[]) => void
+  options: OptionType[]
+  placeholder?: string
+  selectInputWidth?: number
+  size?: 'small' | 'medium' | 'large'
+  'data-testid'?: string
+  notice?: string
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:151]: Notice has been replace by helper you can remove it

File: packages/ui/src/components/UnitInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+              {label}
+            </Text>
+          ) : null}
+          {required && label ? (
+            <Icon name="asterisk" sentiment="danger" size={8} />
+          ) : null}
+          {labelInformation && label ? labelInformation : null}
+        </Stack>
+      ) : null}
+      <UnitInputWrapper
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:239]: There is too many borders 

Screenshot 2024-04-30 at 09 58 05

It should have one unique border + a border separation between both components

File: packages/ui/src/components/UnitInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        data-size={size}
+        data-focus={hasFocus}
+        width={width}
+        id={id}
+        data-success={!!success}
+        data-error={!!error}
+        onMouseEnter={() =>
+          !disabled && !readOnly ? setHovered(true) : undefined
+        }
+        onMouseLeave={() => setHovered(false)}
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:251]: This you don't need I believe you can do that with full css, for instance you could say in a parent container:

```css
&:hover {
  ${UnitInputWrapper} {
     ....
   }
}
```

File: packages/ui/src/components/UnitInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+            min={minValue}
+            step={step}
+            data-success={success}
+            data-error={error}
+            data-testid="unit-input"
+            required={required}
+            className={className}
+            onFocus={() =>
+              !disabled && !readOnly ? setHasFocus(true) : undefined
+            }
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:289]: Same goes for focus, this shouldn't be done in JS but in CSS with a parent container you should be able to manage all

File: packages/ui/src/components/UnitInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+          />
+          {error ? <Icon name="alert" sentiment="danger" /> : null}
+          {success && !error ? (
+            <Icon name="checkbox-circle-outline" sentiment="success" />
+          ) : null}
+        </StyledInputWrapper>
+        <CustomSelectInput
+          width={selectInputWidth}
+          id={`${name}-unit`}
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:298]: The id should be based using `id` provided in the parent if it is provided:

```tsx
id={id ? `${id}-unit` : undefined}
```

And the StyledInput too should have an id then

File: packages/ui/src/components/UnitInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+  'data-testid'?: string
+  notice?: string
+  required?: boolean
+  helper?: string
+  unitError?: string
+  width?: ComponentProps<typeof Stack>['width']
+  placeholderUnit?: string
+  autoFocus?: boolean
+  error?: boolean
+  success?: boolean
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:159]: Should be a `string | boolean` in order to follow what we did with other inputs

File: packages/ui/src/components/UnitInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+          }
+          readOnly={readOnly}
+          onBlur={() => setHasFocus(false)}
+        />
+      </UnitInputWrapper>
+      {helper ? (
+        <Text as="p" variant="caption" sentiment={sentiment}>
+          {helper}
+        </Text>
+      ) : null}
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:325]: And so here you can also follow what has been done with other inputs and success, error props

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/UnitInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+
+  useEffect(() => {
+    const selectBar = document.getElementById('select-bar')
+
+    if (selectBar) {
+      selectBar.style.backgroundColor = 'transparent'
+      selectBar.style.border = 'none'
+      selectBar.style.borderRadius = '0px'
+    }
+  }, [])
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:218]: What is this for ? 😅

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/SelectInputV2/index.tsx [hunk 11]
```diff
@@ -204,20 +204,24 @@ export const SelectInputV2 = <IsMulti extends undefined | boolean>({
...
-              ) : null}
-              {required && label ? (
-                <Icon name="asterisk" sentiment="danger" size={8} />
-              ) : null}
-              {labelDescription ?? null}
-            </Stack>
+            {label ? (
+              <Stack direction="row" gap={0.5}>
+                {label ? (
+                  <Text
```
Comment (matthprost) [inline packages/ui/src/components/SelectInputV2/index.tsx:210]: Label already exists in this context you don't need to recheck it

File: packages/ui/src/components/SelectInputV2/index.tsx [hunk 12]
```diff
@@ -204,20 +204,24 @@ export const SelectInputV2 = <IsMulti extends undefined | boolean>({
...
+                    as="label"
+                    variant={
+                      size === 'large' ? 'bodyStrong' : 'bodySmallStrong'
+                    }
+                  >
+                    {label}
+                  </Text>
+                ) : null}
+                {required && label ? (
```
Comment (matthprost) [inline packages/ui/src/components/SelectInputV2/index.tsx:219]: Same here you can only check if required exists

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/UnitInput/__stories__/Playground.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import type { Decorator } from '@storybook/react'
+import { Template } from './Template.stories'
+
+export const Playground = Template.bind({})
+
+Playground.decorators = [
+  Story => (
+    <div>
+      <div style={{ display: 'inline-flex' }}>
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/__stories__/Playground.stories.tsx:9]: Can you add a `height: '200px'` or more in order to see correctly the select input part in storybook?

Screenshot 2024-05-02 at 10 01 10

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/UnitInput/index.tsx [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+  | 'readOnly'
+  | 'required'
+  | 'autoFocus'
+  | 'onKeyDown'
+>
+
+export const UnitInput = ({
+  id,
+  name = '',
+  max = 99999,
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:190]: Any specific reason to limit max to 99999? Might be better to go with `Number.MAX_SAFE_INTEGER` as we do it in NumberInputV2

File: packages/ui/src/components/UnitInput/index.tsx [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+  min = 1,
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:191]: Same here why 1 and not 0?

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/UnitInput/__stories__/State.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+      name="Success"
+    />
+    <UnitInput
+      options={optionsSelect}
+      error="error"
+      label="Error"
+      name="Error"
+    />
+  </Stack>
+)
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/__stories__/State.stories.tsx:35]: You need to get the props and spread it in components or storybook won't add code snippet for some reasons:
Screenshot 2024-05-13 at 15 48 27

```suggestion
export const States: StoryFn<typeof UnitInput> = (props) => (
  <Stack gap={2}>
    <UnitInput
      options={optionsSelect}
      success="success"
      label="Success"
      name="Success"
      {...props}
    />
    <UnitInput
      options={optionsSelect}
      error="error"
      label="Error"
      name="Error"
      {...props}
    />
  </Stack>
)
```

File: packages/ui/src/components/UnitInput/__stories__/Disabled.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+  {
+    label: 'Days',
+    value: 'days',
+  },
+  {
+    label: 'Months',
+    value: 'months',
+  },
+]
+export const Disabled: StoryFn<typeof UnitInput> = () => (
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/__stories__/Disabled.stories.tsx:20]: Same here you need to spread props

**Review by matthprost — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/UnitInput/index.tsx [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+  autoFocus = false,
+  size = 'large',
+  placeholder = '0',
+  placeholderUnit = 'Select item',
+  onChange,
+  onChangeUnitValue,
+  value,
+  unitValue,
+  selectInputWidth = 200,
+  disabled = false,
```
Comment (matthprost) [inline packages/ui/src/components/UnitInput/index.tsx:201]: There is an issue with disabled, seems like mouse pointer is not disabled when hovering number input part but it is when hovering select input part

**Review by matthprost — CHANGES_REQUESTED** [review]

**Review by matthprost — APPROVED** [review]
Except those last feedbacks everything is fine for me congrats! 🚀

File: packages/ui/src/components/UnitInput/index.tsx [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+
+  useEffect(() => {
+    if (onChange && value) {
+      onChange(value)
+    }
+    if (onChangeUnitValue && unitValue) {
+      onChangeUnitValue(unitValue)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
[the remark this answers is not in the document]
Comment (lisalupi) [reply packages/ui/src/components/UnitInput/index.tsx:259]: forgot to delete it ^^

**Review by mhervouet — APPROVED** [review]

**Pull request merged**